beacon-chain/sync/initial-sync/blocks_fetcher.go

```diff
@@ -293,17 +313,195 @@ func (f *blocksFetcher) fetchBlocksFromPeer(
+	// This assumes blocks are sorted by validROBlocks.
+	highest := bwb[len(bwb)-1].block.Block().Slot()
+	// bwb is sorted by slot, so if the last element is outside the retention window, no blobs are needed.
+	if highest < blobWindowStart {
+		return nil
+	}
+	lowest := lowestSlotNeedsBlob(blobWindowStart, bwb)
+	if lowest == nil {
+		return nil
+	}
+	return &p2ppb.BlobSidecarsByRangeRequest{
+		StartSlot: *lowest,
+		Count:     uint64(highest.SubSlot(*lowest)) + 1,
+	}
+}
+
+func lowestSlotNeedsBlob(retentionStart primitives.Slot, bwb []BlockWithVerifiedBlobs) *primitives.Slot {
+	i := sort.Search(len(bwb), func(i int) bool {
+		return bwb[i].block.Block().Slot() >= retentionStart
+	})
+	if i >= len(bwb) {
+		return nil
+	}
+	s := bwb[i].block.Block().Slot()
+	return &s
+}
+
+func sortBlobs(blobs []*p2ppb.BlobSidecar) []*p2ppb.BlobSidecar {
+	sort.Slice(blobs, func(i, j int) bool {
+		if blobs[i].Slot == blobs[j].Slot {
+			return blobs[i].Index < blobs[j].Index
+		}
+		return blobs[i].Slot < blobs[j].Slot
+	})
+
+	return blobs
+}
+
+var errBlobVerification = errors.New("peer unable to serve aligned BlobSidecarsByRange and BeaconBlockSidecarsByRange responses")
+var errMissingBlobsForBlockCommitments = errors.Wrap(errBlobVerification, "blobs unavailable for processing block with kzg commitments")
+var errMismatchedBlobBlockRoot = errors.Wrap(errBlobVerification, "BlockRoot in BlobSidecar does not match the expected root")
+var errMissingBlobIndex = errors.Wrap(errBlobVerification, "missing expected blob index")
+var errMismatchedBlobCommitments = errors.Wrap(errBlobVerification, "commitments at given slot, root and index do not match")
+
+func verifyAndPopulateBlobs(bwb []BlockWithVerifiedBlobs, blobs []*p2ppb.BlobSidecar, blobWindowStart primitives.Slot) ([]BlockWithVerifiedBlobs, error) {
+	// Assumes bwb has already been sorted by validROBlocks.
+	blobs = sortBlobs(blobs)
+	blobi := 0
+	// Loop over all blocks, and each time a commitment is observed, advance the index into the blob slice.
+	// The assumption is that the blob slice contains a value for every commitment in the blocks it is based on,
+	// correctly ordered by slot and blob index.
+	for i, bb := range bwb {
+		block := bb.block.Block()
+		if block.Slot() < blobWindowStart {
+			continue
+		}
+		commits, err := block.Body().BlobKzgCommitments()
+		if err != nil {
+			if errors.Is(err, consensus_types.ErrUnsupportedField) {
+				log.
+					WithField("block_slot", block.Slot()).
+					WithField("retention_start", blobWindowStart).
+					Warn("block with slot within blob retention period has version which does not support commitments")
+				continue
+			}
+			return nil, err
+		}
+		bb.blobs = make([]*p2ppb.BlobSidecar, len(commits))
+		for ci := range commits {
+			// There are more expected commitments in this block, but we've run out of blobs from the response
+			// (out-of-bound error guard).
+			if blobi == len(blobs) {
```

Comment:
What if we get more blobs than there are commitments in the block? Assuming that's even possible.

Reply:
It would throw the alignment off for the next block, so the roots wouldn't match, which would result in `errMismatchedBlobBlockRoot`. Unless it was the last index of the last block in the list, in which case we wouldn't notice it.